src/paillier.rs

```diff
@@ -0,0 +1,154 @@
+use halo2_base::{
+    halo2_proofs::{circuit::Value, plonk::Error},
+    utils::BigPrimeField,
+    Context,
+};
+use num_bigint::BigUint;
+
+use crate::big_uint::{chip::BigUintChip, AssignedBigUint, Fresh, RefreshAux};
+
+#[derive(Debug, Clone)]
+pub struct PaillierChip<'a, F: BigPrimeField> {
+    pub biguint: &'a BigUintChip<'a, F>,
+    pub enc_bits: usize,
+    pub n: &'a BigUint,
+    pub g: &'a BigUint,
+}
+
+impl<'a, F: BigPrimeField> PaillierChip<'a, F> {
+    pub fn construct(
+        biguint: &'a BigUintChip<'a, F>,
+        enc_bits: usize,
+        n: &'a BigUint,
+        g: &'a BigUint,
+    ) -> Self {
+        Self {
+            biguint,
+            enc_bits,
+            n,
+            g,
+        }
+    }
+
+    pub fn encrypt(
+        &self,
+        ctx: &mut Context<F>,
+        m: &BigUint,
+        r: &BigUint,
+    ) -> Result<AssignedBigUint<F, Fresh>, Error> {
+        let n_assigned =
+            self.biguint
+                .assign_integer(ctx, Value::known(self.n.clone()), self.enc_bits)?;
+        let mut g_assigned =
+            self.biguint
+                .assign_integer(ctx, Value::known(self.g.clone()), self.enc_bits)?;
+        let mut r_assigned =
+            self.biguint
+                .assign_integer(ctx, Value::known(r.clone()), self.enc_bits)?;
+
+        let n2 = self.biguint.square(ctx, &n_assigned)?;
+        let aux = RefreshAux::new(
+            self.biguint.limb_bits,
+            n_assigned.num_limbs(),
+            n_assigned.num_limbs(),
+        );
+        let n2 = self.biguint.refresh(ctx, &n2, &aux)?;
+
+        let zero_value = ctx.load_zero();
+
+        g_assigned = g_assigned.extend_limbs(n2.num_limbs() - g_assigned.num_limbs(), zero_value);
+        let gm = self.biguint.pow_mod_fixed_exp(ctx, &g_assigned, m, &n2)?;
+
+        r_assigned = r_assigned.extend_limbs(n2.num_limbs() - r_assigned.num_limbs(), zero_value);
+        let rn = self
+            .biguint
+            .pow_mod_fixed_exp(ctx, &r_assigned, self.n, &n2)?;
+
+        let c = self.biguint.mul_mod(ctx, &gm, &rn, &n2)?;
+
+        Ok(c)
+    }
+}
+
+pub fn paillier_enc(n: &BigUint, g: &BigUint, m: &BigUint, r: &BigUint) -> BigUint {
```

Comment:
Should this be public?

Reply:
This is used in the bench file as well. Making this `pub(crate)`